[jessica2828] iP

jessica2828 opened the pull request:
## BobPrime

> "Your mind is for having ideas, not holding the,." - David Allen, ([source](https://dansilvestre.com/productivity-quotes/))

**Bob** frees your mind of having to remember things you need to do! It is:

- Text-based
- Easy to learn
- ~~FAST~~ _SUPER FAST_ to use

All you need to do is:

1. Download Bob from [here](https://github.com/jessica2828/ip).
2. Double-click it.
3. Add your tasks.
4. Let it manage your tasks for you! 😆 

And it is ***FREE***!

### Features:

- [x] Managing tasks
- [ ] Managing deadlines _(coming soon)_
- [ ] Reminders _(coming soon)_

---

If you are a Java programmer, you can use it to practice Java too. Here's the `main `method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

jessica2828 committed:
Rename main class and file

jessica2828 committed:
Add greeting and exit

jessica2828 committed:
Add support for echo

jessica2828 committed:
Add support for add

jessica2828 committed:
Add support for list

jessica2828 committed:
Add task class

jessica2828 committed:
Add support for printing tasks with status

jessica2828 committed:
Add support for mark, unmark

jessica2828 committed:
Fix bug invalid task index

jessica2828 committed:
Change to switch statements logic

jessica2828 committed:
Add toString

jessica2828 committed:
Add task types: Deadline, Event, Todo

jessica2828 committed:
Add support for adding deadline, todo and event tasks

jessica2828 committed:
Fix bug (tabs in messages) to allow testing to work

jessica2828 committed:
Add support for Automated Text UI Testing

jessica2828 committed:
Add exception class

jessica2828 committed:
Add support for exception handling

jessica2828 committed:
Add support for delete

jessica2828 committed:
Add logo

jessica2828 committed:
Add packages

jessica2828 committed:
Add storage class

jessica2828 committed:
Add support for save and load

jessica2828 committed:
Add support for storage

jessica2828 committed:
Add support for dates and time

jessica2828 committed:
Fix bug on date and time handling

jessica2828 committed:
Add printline

jessica2828 committed:
Add support for date and time

jessica2828 committed:
Merge branch 'branch-Level-8'

jessica2828 committed:
Merge branch 'master' of https://github.com/jessica2828/ip

* 'master' of https://github.com/jessica2828/ip:

jessica2828 committed:
Fix merge conflicts

jessica2828 committed:
Refactor to adhere to OOP, move to packages

jessica2828 committed:
Merger branch-A-MoreOOP

jessica2828 committed:
Refactor to adhere to OOP

jessica2828 committed:
Refactor to adhere to OOP

jessica2828 committed:
Merge branch 'add-gradle-support'

* add-gradle-support:
  build.gradle: Prevent generating a second JAR file
  Bump gradle and lib version
  Add Gradle support

jessica2828 committed:
Add JUnit tests

jessica2828 committed:
Add support for JAR file

jessica2828 committed:
Add Javadoc

jessica2828 committed:
Refactor Command to an abstract class

jessica2828 committed:
Refactor Command to be an abstract class

jessica2828 committed:
Refactor import statements to follow coding standard, remove unused imports

jessica2828 committed:
Add support for find

jessica2828 committed:
Add support for find, fixed bug

jessica2828 committed:
Merge branch 'branch-A-CodingStandard'

jessica2828 committed:
Merge branch 'branch-Level-9'

gavinsoh commented on src/main/java/Bob/Command/DeadlineCommand.java:
```diff
@@ -0,0 +1,42 @@
+    }
+
+    @Override
+    public void execute(ArrayList<Task> tasks, Storage storage, Ui ui) throws BobException {
+        if (taskDescription.isEmpty()) {
+            throw new BobException("Missing deadline description :(");
+        }
+        String[] dlParts = taskDescription.split(" /by ");
+        if (dlParts.length < 2) {
+            throw new BobException("Missing details :(\nPlease use this format: 'deadline [description] /by [dd/MM/yyyy HHmm]' ");
+        }
+        try {
+            DateTimeFormatter formatter = DateTimeFormatter.ofPattern("dd/MM/yyyy HHmm");
+            LocalDateTime deadlineDateTime = LocalDateTime.parse(dlParts[1], formatter);
+
+            Task deadline = new Deadline(dlParts[0], deadlineDateTime);
+            tasks.add(deadline);
+            storage.save(tasks);
+            ui.showAddedTask(deadline, tasks.size());
+        } catch (DateTimeParseException e) {
+            throw new BobException("Invalid date format! Please use dd/MM/yyyy HHmm :(");
+        }
+    }
```
I like how your Exceptions provide the correct input format to the user.

gavinsoh commented on src/main/java/Bob/Command/UnmarkCommand.java:
```diff
@@ -0,0 +1,26 @@
+
+public class UnmarkCommand extends Command {
+    private final int taskIndexUnmark;
+
+    public UnmarkCommand(int taskIndexUnmark) {
+        this.taskIndexUnmark = taskIndexUnmark;
+    }
+
```
Perhaps you could just use 'index' instead of 'taskIndexUnmark'

gavinsoh commented on src/main/java/Bob/Ui/Ui.java:
```diff
@@ -0,0 +1,116 @@
+public class Ui {
+
+    /**
+     * Displays welcome message and logo when the chatbot starts.
+     */
+    public void showWelcome() {
+        String logo = " ____        ____\n"
+                + "| __ )  ___ | __ )\n"
+                + "|  _ \\ / _ \\|  _ \\\n"
+                + "| |_) | (_) | |_) |\n"
+                + "|____/ \\___/|____/\n";
+        printLine();
+        System.out.println("Hello! I'm Bob!\nHow can I help you today?");
+        System.out.println(logo);
+        printLine();
+    }
+
```
I like how you've implemented this logo from the original Duke Chatbot.

jessica2828 committed:
Add support to prepare for GUI implementation

jessica2828 committed:
Add support for GUI implementation returning a string for each command

jessica2828 committed:
Add support for GUI (final)

jessica2828 committed:
Fix: remove unused logic

jessica2828 committed:
Fix GUI implementation

psh12320 commented on data/test_bob.csv:
```diff
@@ -0,0 +1 @@
+T | 0 | Task 2
```
I like that you used a `.csv` file instead of a `.txt' file for storage 👍

engyuhan replied:
Should the /data directory be pushed onto the remote repo?

psh12320 commented on src/main/java/Bob/Bob.java:
```diff
@@ -0,0 +1,167 @@
+
+    // Constructor without argument
+    public Bob() {
+        this("./data/bob.csv");
```
Perhaps consider extracting `"./data/bob/csv"` into a static constant since for easier maintenance?

psh12320 commented on src/main/java/Bob/Bob.java:
```diff
@@ -0,0 +1,167 @@
+     * Starts Bob chatbot, handling user input and executing commands.
+     * This method is primarily for console-based interaction.
+     */
+    public void run() {
```
I like this clean bit of code!

psh12320 commented on src/main/java/Bob/Command/FindCommand.java:
```diff
@@ -0,0 +1,30 @@
+
+    @Override
+    public String execute(ArrayList<Task> tasks, Storage storage, Ui ui) throws BobException {
+        ArrayList<Task> matchingTasks = tasks.stream()
```
I like this bit of code. Nicely done using streams 👍

psh12320 commented on src/main/java/Bob/Storage/Storage.java:
```diff
@@ -0,0 +1,100 @@
+
+public class Storage {
+    private String filePath;
+    private static final DateTimeFormatter DATE_TIME_FORMATTER = DateTimeFormatter.ofPattern("dd/MM/yyyy HHmm");
```
I like the use of SCREAMING_SNAKE_CASE here 👍

psh12320 reviewed (commented):
Overall LGTM. Just a few nits to fix.

engyuhan commented on src/main/java/Bob/Bob.java:
```diff
@@ -0,0 +1,167 @@
+//
+//import java.util.Scanner;
+//
+///**
```
Any reason why you kept this dead code in the file? I see this in some other places as well.

engyuhan commented on src/main/java/Bob/Storage/Storage.java:
```diff
@@ -0,0 +1,100 @@
+                        task = new Todo(parts[2]);
+                        break;
+                    case "D":
+                        String dateString = parts[3];
```
Should this be extracted out as a separate method?

engyuhan commented on src/main/resources/images/bob.png:
nice

engyuhan reviewed (commented):
LGTM, just a few nits to pick.

jessica2828 committed:
Add assert feature

jessica2828 committed:
Change package names to lowercase

jessica2828 committed:
Change import statements to follow new package names

jessica2828 committed:
Refactor packages to follow coding standard

jessica2828 committed:
Storage class: extract file handling methods

The Storage.load() method handles multiple tasks including file setup,
reading tasks, and creating different task types; leading to a method that
is hard to follow and maintain.

The method's multiple level of abstraction violates SLAP principle and
makes the method more complex than necessary, it is harder to read and
increases the chances of errors when modifying the code.

Let's extract methods for file setup and task parsing into their own
separate methods. Break down the load method into initializeFile(),
loadCurrentTasks(), parseTask(), createTaskFromParts(), and parseDeadline().

Separating these responsibilities improves readability, by ensuring
each method does one thing. Adherence to SLAP makes the code easier to
maintain, modify, and debug.

jessica2828 committed:
Refactor classes to improve code quality

jessica2828 committed:
Parser class: add assertions

Parser.parse() method handles various commands without defining key
assumptions, which can lead to bugs or incorrect behavior during runtime
if invalid inputs are passed in.

Without assertion, it may beharder to catch potential issues early,
increasing the likelihood of errors.

Let's add assert statements to verify key assumptions, such as non-null
inputs, non-empty task descriptions, and valid task indices.

Using assertions provides a lightweight way to enforce and document
assumptions without affecting performance in production, as assertions
can be disabled. This approach improves code clarity and reduces the
risk of subtle bugs during the development phase.

jessica2828 committed:
Merge pull request #2 from jessica2828/branch-A-CodeQuality

Improve code quality

jessica2828 committed:
Merge master

jessica2828 committed:
Merge branch 'master' into branch-A-Assertions

* master:
  Refactor classes to improve code quality
  Storage class: extract file handling methods
  Refactor packages to follow coding standard
  Change import statements to follow new package names

jessica2828 committed:
Merge pull request #1 from jessica2828/branch-A-Assertions

Add assertions

jessica2828 committed:
Add support for C-Update extension

jessica2828 committed:
Update javadocs

jessica2828 committed:
Improve GUI

jessica2828 committed:
Update the User Guide

jessica2828 committed:
Remove old README file

jessica2828 committed:
Update User Guide

jessica2828 committed:
Make changes to user guide
